src/libos/crates/page-cache/src/page_cache.rs

```diff
@@ -0,0 +1,296 @@
+    marker: PhantomData<(K, A)>,
+}
+
+/// Page cache flusher.
+///
+/// A page cache must be equiped with a user-given
+/// flusher `F: PageCacheFlusher` so that when
+/// the memory is low, the page cache mechanism
+/// can automatically flush dirty pages and
+/// subsequently evict pages.
+///
+/// This trait has only one method.
+#[async_trait]
+pub trait PageCacheFlusher: Send + Sync {
+    /// Flush the dirty pages in a page cache.
+    ///
+    /// If success, then the return value is
+    /// the number of dirty pages that are flushed.
+    async fn flush(&self) -> Result<usize>;
+}
+
+impl<K: PageKey, A: PageAlloc> PageCache<K, A> {
+    /// Create a new page cache.
+    ///
+    /// Specify a flusher which implements `PageCacheFlusher`.
+    pub fn new(flusher: Arc<dyn PageCacheFlusher>) -> Self {
+        info!("[PageCache] new");
+        let new_self = Self(Arc::new(PageCacheInner::new(flusher)));
+        PageEvictor::<K, A>::register(&new_self);
+        new_self
+    }
+
+    /// Acquire the page that corresponds to the key.
+    ///
+    /// Return `None` if there are no available pages.
+    /// In this case, the user can use the
+    /// `poll` method to wait for the readiness of the
+    /// page cache.
+    pub fn acquire(&self, key: K) -> Option<PageHandle<K, A>> {
+        let mut cache = self.0.cache.lock();
+        // Cache hit
+        if let Some(page_handle_incache) = cache.get(&key.into()) {
+            return Some(page_handle_incache.clone());
+        // Cache miss
+        } else {
+            // Cache miss and a new page is allocated
+            if let Some(page_handle) = PageHandle::new(key) {
+                cache.put(key.into(), page_handle.clone());
+                return Some(page_handle);
+            }
+        }
+        // Cache miss and no free space for new page
+        None
+    }
+
+    /// Release the page.
+    ///
+    /// All page handles obtained via the `acquire` method
+    /// must be returned via the `release` method.
+    pub fn release(&self, page_handle: PageHandle<K, A>) {
+        // The dirty_set traces dirty pages in order
+        let mut dirty_set = self.0.dirty_set.lock();
+        let page_guard = page_handle.lock();
+        // Update dirty_set when page_handle released
+        if page_guard.state() == PageState::Dirty {
+            dirty_set.insert(page_handle.key().into());
+        } else {
+            dirty_set.remove(&page_handle.key().into());
+        }
+    }
+
+    /// Pop a number of dirty pages and switch their state to
+    /// "Flushing".
+    ///
+    /// The handles of dirty pages are pushed into the given `Vec`.
+    /// The dirty page ids are in ascending order.
+    /// Return flush numbers.
+    pub fn flush_dirty(&self, dirty: &mut Vec<PageHandle<K, A>>) -> usize {
+        let cache = self.0.cache.lock();
+        // The dirty_set traces dirty pages in order
+        let mut dirty_set = self.0.dirty_set.lock();
+        let mut flush_num = 0;
```

Comment:
I think better not because the user cannot be aware of the number of dirty pages all the time. Typically they may keep calling `pop_dirty_to_flush()` until the `dirty_set` is empty.